### ISSUE TITLE
Built-in named arguments

### ISSUE THREAD

**Issue by pvider** [marcobambini/gravity#197]
By 'built-in' I mean that argument names for all functions should be simply generated from their definitions, for example: `func f(x, y, z) {.....}` would allow positional arguments call `f(10, 20, 30)`, named `f(y = 20, z = 30, x = 10)` and mixed `f(x = 10, 20, 30)`. Here every name defines exact position of the named argument to mix it with positional arguments. In mixed case the positional arguments should be counted from the position number of  named argument before it. If a name of used named argument does not exist then Gravity should rise an error.

**Comment by marcobambini** [comment]
@pvider I really like the idea but it is not so easy to implement.
It should be a static decision (the compiler should generate the right call based on the information provided by the source code) but gravity is so dynamic that I am blocked on this case:

```swift
func test1(a, b) {
    return a + b;
}

func test2(c,d) {
    return c * d;
}

func test() {
    if (is_sunday()) return test1;
    else return test2;
}

func main() {
    return test()(a:10, b:20);
}
```

This code would work ONLY on Sunday and it would be a runtime decision, that means that how to pass parameters should be handled at runtime and that would add a lot of overhead in function/method call.

Am I missing something?

**Comment by pvider** [comment]
No, you are totally right, it is not always possible to do static decision.  We can use different things to convert argument names to positions at runtime: Gravity maps, string array, sort of preoptimized binary search tree. I like string array because it has lowest memory overhead, easy implementation and strcmp function is optimized enough to provide great speed for small number of arguments. However it has O(n) element search complexity which can be lowered to O(1) if users will use names in the same order as  function declaration does.

**Comment by marcobambini** [comment]
@pvider as I said I really like the idea to have named arguments but I don't like the idea to slowdown every call. I need to find out a way to balance pros and cons. I think that the OPCODE used by the VM for the call should be different from the OPCODE used for a named args call, in this way the slowdown will affect only the named arguments call. strcmp is not an issue inside Gravity because all strings are hashed and most of the time the overhead of the comparison is just the time required to compare two numbers.

Do you know how other languages solved this problem?

**Comment by pvider** [comment]
**@marcobambini**, Other dynamic languages solve this by using hash-tables. Lua directly uses hash-tables with some sugar (it allows omitting `()` in `({...})`) which is ineffecient since every time it creates non-optimized tables for every call. Python uses 3 instructions to call functions (CALL_FUNCTION, CALL_FUNCTION_KW, CALL_FUNCTION_EX). As I understand Python also uses optimized hash-tables (it has BUILD_CONST_KEY_MAP opcode).
I agree we should use different opcodes to call functions with named arguments. And we should use binary search trees optimized for constants. How about this: each defined function stores _sorted_ string array (the keys) and integer array (the positions, since keys are sorted), each function call  with named arguments loads _sorted_ string array (keys of arguments). Then we need to map strings to positions with binary search: since everything is sorted for each key we start narrowing the range of keys until we get the key what we need (or don't get). If that key exists we have it's index in string array and we use it to get that key position from integer array and finally put the argument into right place. The complexity is O(log2n).

**Comment by sandrobilbeisi** [comment]
a very similar earlier programming language known as **Biferno** supports optional named parameters  https://github.com/Biferno

**Comment by marcobambini** [comment]
My goal in order to implement named arguments was to reuse as much code as possible from the already implemented CALL method (which takes care of a lot of others details).

I am going to add a new opcode NCALL (with the same exact syntax as CALL) with the addition that the keys used in the named call will be stored in the stack. NCALL will be responsible to fix the position of the passed arguments and to raise a runtime error in case of unknown keys.

The remain of the opcode will be equal to CALL, that the reason why they both will share the same code block. The overhead for the normal CALL will be a couple of if statements.
